tests/agent_server/test_server_agent.py

```diff
@@ -3887,3 +3893,225 @@ def test_deploy_response_matrix_invariants():
+    no_agent_backoff: None,
+    caplog,
+    monkeypatch,
+):
+    """
+    Test goal: make sure that failed resources are correctly logged by the `resource_action` logger.
+    """
+
+    caplog.set_level(logging.INFO)
+    resource_container.Provider.reset()
+    agent_name = "agent1"
+    myagent_instance = agent._instances[agent_name]
+
+    resource_container.Provider.set("agent1", "key1", "value1")
+
+    def get_resources(version, value_resource_three):
+        return [
+            {
+                "key": "key1",
+                "value": "value2",
+                "id": "test::Resource[agent1,key=key1],v=%d" % version,
+                "send_event": False,
+                "purged": False,
+                "requires": [],
+            },
+            {
+                "key": "key2",
+                "value": "value2",
+                "id": "test::AgentConfig[agent1,key=key2],v=%d" % version,
+                "send_event": False,
+                "purged": False,
+                "requires": ["test::Resource[agent1,key=key1],v=%d" % version],
+            },
+            {
+                "key": "key3",
+                "value": value_resource_three,
+                "id": "test::Resource[agent1,key=key3],v=%d" % version,
+                "send_event": False,
+                "purged": False,
+                "requires": ["test::AgentConfig[agent1,key=key2],v=%d" % version],
+            },
+        ]
+
+    version1 = await clienthelper.get_version()
+
+    # Initial deploy
+    await _deploy_resources(client, environment, get_resources(version1, "value2"), version1, push=False)
+
+    async def ensure_code(code: Collection[ResourceInstallSpec]) -> executor.FailedResources:
+        raise RuntimeError(f"Failed to install handler `test` version={version1}")
+
+    monkeypatch.setattr(myagent_instance.executor_manager.process, "ensure_code", ensure_code)
+
+    await myagent_instance.get_latest_version_for_agent(
+        DeployRequest(reason="Deploy 1", is_full_deploy=False, is_periodic=False)
+    )
+
+    monkeypatch.undo()
+
+    module_folder = pathlib.Path(__file__).resolve().parent.parent.parent
+
+    expected_error_message = f"""multiple resources: All resources of type `test::Resource` failed to load handler code or \
+install handler code dependencies: `Could not set up executor for agent1: Failed to install handler `test` version=1`
+  File "{module_folder}/src/inmanta/agent/agent.py", line 854, in setup_executor
+    current_executor = await self.executor_manager.get_executor(self.name, self.uri, code)
+                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  File "{module_folder}/src/inmanta/agent/in_process_executor.py", line 502, in get_executor
+    out.failed_resources = await self.process.ensure_code(code)
+                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  File "{module_folder}/tests/agent_server/test_server_agent.py", line 3953, in ensure_code
+"""
```

Comment:
This is a bit too strict to be used in an assertion, because it contains line numbers. As such, the test case will break every time the line number changes. It's sufficient to verify that the exception comes from the location where it was raised. So the following pattern should match: `File "{module_folder}/tests/agent_server/test_server_agent.py", line \d+, in ensure_code`. All other parts of the stack trace are not relevant.